File: orgabbrev.py
import re
import unicodedata

class CompanyAbbrev:
    LEGAL_SUFFIXES = {
        "en": ["inc", "llc", "corp", "ltd", "llp", "plc", "corporation", "limited", "company"],
        "fr": ["sa", "sarl", "sas"],
        "es": ["sa", "sl", "sau", "slne"],
        "pt": ["lda", "ltda", "sa", "me"],
        "de": ["gmbh", "ag", "kg", "ug"],
        "it": ["srl", "spa", "snc", "sas"],
        "ru": ["ооо", "зао", "оао", "ao"],
        "zh": ["有限公司", "集团", "公司"],
        "ja": ["株式会社", "有限会社"],
        "ko": ["유한회사", "주식회사"],
        "he": ["בעמ", "חברה בעמ"],
        "ar": ["شذمم", "ش.م.م", "مؤسسة", "شركة"]
    }

    NOISE_WORDS = {
        "en": ["the", "of", "and"],
        "fr": ["le", "la", "les", "de", "et"],
        "es": ["el", "la", "los", "de", "y"],
        "pt": ["o", "a", "os", "as", "de", "e"],
        "de": ["der", "die", "das", "und", "von"],
        "it": ["il", "la", "lo", "gli", "dei", "e"],
        "ru": ["и", "из", "в"],
        "zh": [],
        "ja": [],
        "ko": [],
        "he": ["של", "ו"],
        "ar": ["و", "من", "ال"]
    }

    @staticmethod
    def normalize_nfd(text: str) -> str:
        return unicodedata.normalize("NFD", text)

    @staticmethod
    def extract_existing_acronym(name: str):
        match = re.search(r"\(([A-Z]{2,6})\)\s*$", name)
        if not match:
            return name.strip(), None
        acronym = match.group(1)
        first_letter_match = re.search(r"[A-Za-z]", name)
        if not first_letter_match or acronym[0] != first_letter_match.group(0).upper():
            return name.strip(), None
        name = re.sub(r"\s*\([A-Z]{2,6}\)\s*$", "", name)
        return name.strip(), acronym

    @staticmethod
    def normalize_punctuation(name: str) -> str:
        # Keep letters, digits, spaces, &, hyphen; replace other punctuation with space
        name = re.sub(r"[^\w\s&-]+", " ", name)
        name = re.sub(r"_", " ", name)
        name = re.sub(r"\s+", " ", name)
        return name.strip()

    @classmethod
    def remove_legal_suffix(cls, name: str, lang: str):
        # Normalize name for suffix removal
        normalized = cls.normalize_punctuation(name).lower()
        for suffix in cls.LEGAL_SUFFIXES.get(lang, []):
            pattern = re.compile(r"\b" + re.escape(suffix.lower()) + r"$")
            if pattern.search(normalized):
                # Remove suffix from the normalized string
                name = re.sub(r"\b" + re.escape(suffix) + r"$", "", name, flags=re.IGNORECASE).strip()
                break
        return name

    @classmethod
    def clean_words(cls, name: str, lang: str):
        noise = set(cls.NOISE_WORDS.get(lang, []))
        words = re.split(r"[ \-&]+", name)
        cleaned = [w for w in words if w and w.lower() not in noise]
        return " ".join(cleaned)

    @staticmethod
    def propose_acronym(name: str, word_count_threshold=3):
        words = re.split(r"[ \-&]+", name)
        if len(words) < word_count_threshold:
            return None
        if any(re.fullmatch(r"[A-Z]{2,}", w) for w in words):
            return None
        return "".join(w[0].upper() for w in words if w != "&")

    @classmethod
    def abbreviate(cls, name: str, lang: str = "en", propose_acronym: bool = True):
        if not name:
            return ""

        # 1. Extract parenthesized acronym if valid
        name, existing_acronym = cls.extract_existing_acronym(name)

        # 2. Normalize Unicode and punctuation
        name = cls.normalize_nfd(name)
        name = cls.normalize_punctuation(name)

        # 3. Remove legal suffix
        name = cls.remove_legal_suffix(name, lang)

        # 4. Clean noise words (keep & for counting)
        name = cls.clean_words(name, lang)

        acronym = existing_acronym
        if not acronym and propose_acronym:
            words_for_count = re.split(r"[ \-&]+", name)
            if len(words_for_count) >= 3:
                acronym = cls.propose_acronym(name)

        # 5. Remove & for final output
        name = name.replace("&", "").replace("  ", " ").strip()

        if acronym:
            return f"{name} ({acronym})"
        return name

# -----------------------
# Demo
# -----------------------
if __name__ == "__main__":
    abbr = CompanyAbbrev()
    examples = [
        ("Hewlett-Packard Company, Inc.", "en"),
        ("International Business Machines Corporation", "en"),
        ("Grupo de Industrias Unidas S.A.", "es"),
        ("Société Générale S.A.", "fr"),
        ("Bayerische Motoren Werke AG", "de"),
        ("株式会社トヨタ自動車", "ja"),
        ("有限责任公司中石化", "zh"),
        ("주식회사 삼성전자", "ko"),
        ("ООО Газпром", "ru"),
        ("חברה בעמ אל על", "he"),
        ("شركة الاتصالات السعودية", "ar"),
    ]

    for name, lang in examples:
        result = abbr.abbreviate(name, lang)
        print(f"[{lang}] {name:40} → {result}")

File: orgabbrev.js
class CompanyAbbrev {
    static LEGAL_SUFFIXES = {
        en: ["inc", "llc", "corp", "ltd", "llp", "plc", "corporation", "limited", "company"],
        fr: ["sa", "sarl", "sas"],
        es: ["sa", "sl", "sau", "slne"],
        pt: ["lda", "ltda", "sa", "me"],
        de: ["gmbh", "ag", "kg", "ug"],
        it: ["srl", "spa", "snc", "sas"],
        ru: ["ооо", "зао", "оао", "ao"],
        zh: ["有限公司", "集团", "公司"],
        ja: ["株式会社", "有限会社"],
        ko: ["유한회사", "주식회사"],
        he: ["בעמ", "חברה בעמ"],
        ar: ["شذمم", "ش.م.م", "مؤسسة", "شركة"]
    };

    static NOISE_WORDS = {
        en: ["the", "of", "and"],
        fr: ["le", "la", "les", "de", "et"],
        es: ["el", "la", "los", "de", "y"],
        pt: ["o", "a", "os", "as", "de", "e"],
        de: ["der", "die", "das", "und", "von"],
        it: ["il", "la", "lo", "gli", "dei", "e"],
        ru: ["и", "из", "в"],
        zh: [],
        ja: [],
        ko: [],
        he: ["של", "ו"],
        ar: ["و", "من", "ال"]
    };

    static normalizeNFD(text) {
        return text.normalize("NFD");
    }

    static normalizePunctuation(name) {
        name = name.replace(/[^\w\s&-]+/g, " ");
        name = name.replace(/_/g, " ");
        name = name.replace(/\s+/g, " ");
        return name.trim();
    }

    static extractExistingAcronym(name) {
        const match = name.match(/\(([A-Z]{2,6})\)\s*$/);
        if (match) {
            const acronym = match[1];
            const firstLetterMatch = name.match(/[A-Za-z]/);
            if (firstLetterMatch && acronym[0] === firstLetterMatch[0].toUpperCase()) {
                name = name.replace(/\s*\([A-Z]{2,6}\)\s*$/, "").trim();
                return { name, acronym };
            }
        }
        return { name, acronym: null };
    }

    static removeLegalSuffix(name, lang) {
        let normalized = this.normalizePunctuation(name).toLowerCase();
        const suffixes = this.LEGAL_SUFFIXES[lang] || [];
        for (let suffix of suffixes) {
            const pattern = new RegExp("\\b" + suffix + "$", "i");
            if (pattern.test(normalized)) {
                name = name.replace(pattern, "").trim();
                break;
            }
        }
        return name;
    }

    static cleanWords(name, lang) {
        const noise = new Set(this.NOISE_WORDS[lang] || []);
        const words = name.split(/[ \-&]+/);
        const cleaned = words.filter(w => w && !noise.has(w.toLowerCase()));
        return cleaned.join(" ");
    }

    static proposeAcronym(name) {
        const words = name.split(/[ \-&]+/);
        if (words.length < 3) return null;
        if (words.some(w => /^[A-Z]{2,}$/.test(w))) return null;
        return words.filter(w => w !== "&").map(w => w[0].toUpperCase()).join("");
    }

    static abbreviate(name, lang = "en", proposeAcronym = true) {
        if (!name) return "";

        let { name: processedName, acronym: existingAcronym } = this.extractExistingAcronym(name);

        processedName = this.normalizeNFD(processedName);
        processedName = this.normalizePunctuation(processedName);
        processedName = this.removeLegalSuffix(processedName, lang);
        processedName = this.cleanWords(processedName, lang);

        let acronym = existingAcronym;
        if (!acronym && proposeAcronym) {
            const wordsForCount = processedName.split(/[ \-&]+/);
            if (wordsForCount.length >= 3) {
                acronym = this.proposeAcronym(processedName);
            }
        }

        processedName = processedName.replace(/&/g, "").replace(/\s+/g, " ").trim();

        if (acronym) return `${processedName} (${acronym})`;
        return processedName;
    }
}

/* Example usage
console.log(CompanyAbbrev.abbreviate("Hewlett-Packard Company, Inc.")); // Hewlett Packard Company (HPC)
console.log(CompanyAbbrev.abbreviate("International Business Machines Corporation")); // International Business Machines (IBM)
console.log(CompanyAbbrev.abbreviate("Grupo de Industrias Unidas S.A.", "es")); // Grupo Industrias Unidas (GIU)
console.log(CompanyAbbrev.abbreviate("Société Générale S.A.", "fr")); // Société Générale (SG)
console.log(CompanyAbbrev.abbreviate("Bayerische Motoren Werke AG", "de")); // Bayerische Motoren Werke (BMW)
*/